src/main/scala/io/iohk/ethereum/ledger/BranchResolution.scala

```diff
@@ -0,0 +1,77 @@
+package io.iohk.ethereum.ledger
+
+import io.iohk.ethereum.domain.{ Block, BlockHeader, Blockchain }
+
+class BranchResolution(blockchain: Blockchain) {
+
+  def resolveBranch(headers: Seq[BlockHeader]): BranchResolutionResult = {
+    if (!areHeadersFormChain(headers) || headers.last.number < blockchain.getBestBlockNumber()) {
+      InvalidBranch
+    } else {
+      // Dealing with a situation when genesis block is included in the received headers,
+      // which may happen in the early block of private networks
+      val result = for {
+        genesisHeader      <- blockchain.getBlockHeaderByNumber(0)
+        givenHeadOfHeaders <- headers.headOption
+        isGenesisNumber     = givenHeadOfHeaders.number == genesisHeader.number
+        isGenesisHash       = givenHeadOfHeaders.hash == genesisHeader.hash
+        reachedGenesis      = isGenesisHash && isGenesisNumber
+        parentIsKnown       = blockchain.getBlockHeaderByHash(givenHeadOfHeaders.parentHash).isDefined
+      } yield parentIsKnown || reachedGenesis
```

Comment:
If `reachedGenesis` is true you unnecessary call the `blockchain.getBlockHeaderByHash`. You could write this code in different order.